--- a/drivers/clk/qcom/gcc-ipq806x.c	2018-03-21 20:33:18.581000000 +0800
+++ b/drivers/clk/qcom/gcc-ipq806x.c	2018-03-21 22:05:44.838216822 +0800
@@ -2801,6 +2801,81 @@
 	},
 };
 
+static const struct freq_tbl clk_tbl_ce5[] = {
+	{  48000000, P_PLL8, 8,   1,   1 },
+	{ 150000000, P_PLL8, 8,  25,   8 },
+	{ 160000000, P_PLL8, 8,  20,   6 },
+	{ 213200000, P_PLL8, 1,   5,   9 },
+	{ }
+};
+
+static struct clk_rcg ce5_src_clk = {
+	.ns_reg = 0x36c4,
+	.mn = {
+		.mnctr_en_bit = 8,
+		.mnctr_reset_bit = 7,
+		.mnctr_mode_shift = 5,
+		.n_val_shift = 16,
+		.m_val_shift = 16,
+		.width = 16,
+	},
+	.p = {
+		.pre_div_shift = 3,
+		.pre_div_width = 2,
+	},
+	.s = {
+		.src_sel_shift = 0,
+		.parent_map = gcc_pxo_pll8_map,
+	},
+	.freq_tbl = clk_tbl_ce5,
+	.clkr = {
+		.enable_reg = 0x36c4,
+		.enable_mask = BIT(7),
+		.hw.init = &(struct clk_init_data) {
+			.name = "ce5_src_clk",
+			.parent_names = gcc_pxo_pll8,
+			.num_parents = 2,
+			.ops = &clk_rcg_ops,
+		},
+	},
+};
+
+static struct clk_branch ce5_core_clk = {
+	.halt_reg = 0x2fdc,
+	.halt_bit = 5,
+	.clkr = {
+		.enable_reg = 0x36cc,
+		.enable_mask = BIT(4),
+		.hw.init = &(struct clk_init_data) {
+			.name = "ce5_core_clk",
+			.parent_names = (const char *[]) {
+				"ce5_src_clk",
+			},
+			.num_parents = 1,
+			.ops = &clk_branch_ops,
+			.flags = CLK_SET_RATE_PARENT,
+		},
+	},
+};
+
+static struct clk_branch ce5_h_clk = {
+	.halt_reg = 0x3c20,
+	.halt_bit = 11,
+	.clkr = {
+		.enable_reg = 0x3c6c,
+		.enable_mask = BIT(4),
+		.hw.init = &(struct clk_init_data) {
+			.name = "ce5_h_clk",
+			.parent_names = (const char *[]) {
+				"ce5_src_clk",
+			},
+			.num_parents = 1,
+			.ops = &clk_branch_ops,
+			.flags = CLK_SET_RATE_PARENT,
+		},
+	},
+};
+
 static int nss_core_clk_set_rate(struct clk_hw *hw, unsigned long rate,
 				 unsigned long parent_rate)
 {
@@ -3002,6 +3080,9 @@
 	[PLL9] = &hfpll0.clkr,
 	[PLL10] = &hfpll1.clkr,
 	[PLL12] = &hfpll_l2.clkr,
+	[CE5_SRC] = &ce5_src_clk.clkr,
+	[CE5_H_CLK] = &ce5_h_clk.clkr,
+	[CE5_CORE_CLK] = &ce5_core_clk.clkr,
 };
 
 static const struct qcom_reset_map gcc_ipq806x_resets[] = {
@@ -3162,6 +3162,13 @@
	[NSS_CAL_PRBS_RST_N_RESET] = { 0x3b60, 26 },
	[NSS_LCKDT_RST_N_RESET] = { 0x3b60, 27 },
	[NSS_SRDS_N_RESET] = { 0x3b60, 28 },
+	[CE5_H_CLK_RESET] = { 0x3c6c, 7 },
+	[CE5_CORE_CLK_RESET] = { 0x36cc, 7 },
+	[CRYPTO_ENG1_RESET] = { 0x3e00, 0},
+	[CRYPTO_ENG2_RESET] = { 0x3e04, 0},
+	[CRYPTO_ENG3_RESET] = { 0x3e08, 0},
+	[CRYPTO_ENG4_RESET] = { 0x3e0c, 0},
+	[CRYPTO_AHB_RESET] = { 0x3e10, 0},
 };

 static const struct regmap_config gcc_ipq806x_regmap_config = {
--- a/include/dt-bindings/reset/qcom,gcc-ipq806x.h	2018-03-22 22:39:56.224047184 +0800
+++ b/include/dt-bindings/reset/qcom,gcc-ipq806x.h	2018-03-22 22:48:15.912047184 +0800
@@ -171,5 +171,12 @@
 #define NSS_CAL_PRBS_RST_N_RESET			154
 #define NSS_LCKDT_RST_N_RESET				155
 #define NSS_SRDS_N_RESET				156
+#define CE5_H_CLK_RESET					157
+#define CE5_CORE_CLK_RESET				158
+#define CRYPTO_ENG1_RESET				159
+#define CRYPTO_ENG2_RESET				160
+#define CRYPTO_ENG3_RESET				161
+#define CRYPTO_ENG4_RESET				162
+#define CRYPTO_AHB_RESET				163
 
 #endif
