diff --git a/ath10k-5.8/core.c b/ath10k-5.8/core.c
index 592975b..b41c89a 100644
--- a/ath10k-5.8/core.c
+++ b/ath10k-5.8/core.c
@@ -2756,7 +2756,7 @@ static int ath10k_init_uart(struct ath10k *ar)
 
 static int ath10k_init_hw_params(struct ath10k *ar)
 {
-	const struct ath10k_hw_params *uninitialized_var(hw_params);
+	const struct ath10k_hw_params *hw_params;
 	int i;
 
 	for (i = 0; i < ARRAY_SIZE(ath10k_hw_params_list); i++) {
